### ISSUE TITLE
wpa_supplicant does not work with multiple interfaces

### ISSUE THREAD

**Issue by lheckemann** [NixOS/nixpkgs#61391]
## Issue description

If multiple wireless network interfaces are available, `wpa_supplicant` will be invoked incorrectly by its start script, dump a help message to the log, and exit.

Line where this happens:
https://github.com/NixOS/nixpkgs/blob/f5ab71ba28784a56b8b4878914ecd6f384ce666e/nixos/modules/services/networking/wpa_supplicant.nix#L236

<details>
<summary>Unit log</summary>

```
-- Reboot --
May 12 18:12:51 usbdisk systemd[1]: Started WPA Supplicant.
May 12 18:12:51 usbdisk wpa_supplicant[1425]: Successfully initialized wpa_supplicant
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]: wpa_supplicant v2.8
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]: Copyright (c) 2003-2019, Jouni Malinen <j@w1.fi> and contributors
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]: This software may be distributed under the terms of the BSD license.
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]: See README for more details.
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]: This product includes software developed by the OpenSSL Project
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]: for use in the OpenSSL Toolkit (http://www.openssl.org/)
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]: usage:
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   wpa_supplicant [-BddhKLqqstuvW] [-P<pid file>] [-g<global ctrl>] \
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:         [-G<group>] \
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:         -i<ifname> -c<config file> [-C<ctrl>] [-D<driver>] [-p<driver_param>] \
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:         [-b<br_ifname>] [-e<entropy file>] [-f<debug file>] \
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:         [-o<override driver>] [-O<override ctrl>] \
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:         [-N -i<ifname> -c<conf> [-C<ctrl>] [-D<driver>] \
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:         [-m<P2P Device config file>] \
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:         [-p<driver_param>] [-b<br_ifname>] [-I<config file>] ...]
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]: drivers:
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   nl80211 = Linux nl80211/cfg80211
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   wext = Linux wireless extensions (generic)
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   wired = Wired Ethernet driver
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]: options:
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   -b = optional bridge interface name
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   -B = run daemon in the background
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   -c = Configuration file
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   -C = ctrl_interface parameter (only used if -c is not)
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   -d = increase debugging verbosity (-dd even more)
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   -D = driver name (can be multiple drivers: nl80211,wext)
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   -e = entropy file
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   -f = log output to debug file instead of stdout
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   -g = global ctrl_interface
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   -G = global ctrl_interface group
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   -h = show this help text
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   -i = interface name
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   -I = additional configuration file
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   -K = include keys (passwords, etc.) in debug output
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   -L = show license (BSD)
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   -m = Configuration file for the P2P Device interface
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   -N = start describing new interface
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   -o = override driver parameter for new interfaces
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   -O = override ctrl_interface parameter for new interfaces
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   -p = driver parameters
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   -P = PID file
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   -q = decrease debugging verbosity (-qq even less)
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   -s = log output to syslog instead of stdout
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   -t = include timestamp in debug messages
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   -u = enable DBus control interface
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   -v = show version
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   -W = wait for a control interface monitor before starting
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]: example:
May 12 18:12:53 usbdisk ajph60xc285xi286yi5yyc7m5gv6gdf0-unit-script-wpa_supplicant-start[1425]:   wpa_supplicant -Dnl80211 -iwlan0 -c/etc/wpa_supplicant.conf
May 12 18:12:53 usbdisk wpa_supplicant[1425]: nl80211: deinit ifname=wlp0s29f7u1u2 disabled_11b_rates=0
May 12 18:12:53 usbdisk wpa_supplicant[1425]: wlp0s29f7u1u2: CTRL-EVENT-TERMINATING
May 12 18:12:53 usbdisk systemd[1]: wpa_supplicant.service: Main process exited, code=exited, status=255/n/a
May 12 18:12:53 usbdisk systemd[1]: wpa_supplicant.service: Failed with result 'exit-code'.
```
</details>

<details>
<summary>Output of running the unit script manually with `bash -x`</summary>

```
++ cd /sys/class/net
++ echo enp2s0f0 lo tinc.lugnet virbr0 virbr0-nic wlp0s29f7u1u2 wlp3s0b1
+ for i in $(cd /sys/class/net && echo *)
+ DEVTYPE=
+ source /sys/class/net/enp2s0f0/uevent
++ INTERFACE=enp2s0f0
++ IFINDEX=2
+ '[' '' = wlan -o -e /sys/class/net/enp2s0f0/wireless ']'
+ for i in $(cd /sys/class/net && echo *)
+ DEVTYPE=
+ source /sys/class/net/lo/uevent
++ INTERFACE=lo
++ IFINDEX=1
+ '[' '' = wlan -o -e /sys/class/net/lo/wireless ']'
+ for i in $(cd /sys/class/net && echo *)
+ DEVTYPE=
+ source /sys/class/net/tinc.lugnet/uevent
++ INTERFACE=tinc.lugnet
++ IFINDEX=9
+ '[' '' = wlan -o -e /sys/class/net/tinc.lugnet/wireless ']'
+ for i in $(cd /sys/class/net && echo *)
+ DEVTYPE=
+ source /sys/class/net/virbr0/uevent
++ DEVTYPE=bridge
++ INTERFACE=virbr0
++ IFINDEX=6
+ '[' bridge = wlan -o -e /sys/class/net/virbr0/wireless ']'
+ for i in $(cd /sys/class/net && echo *)
+ DEVTYPE=
+ source /sys/class/net/virbr0-nic/uevent
++ INTERFACE=virbr0-nic
++ IFINDEX=7
+ '[' '' = wlan -o -e /sys/class/net/virbr0-nic/wireless ']'
+ for i in $(cd /sys/class/net && echo *)
+ DEVTYPE=
+ source /sys/class/net/wlp0s29f7u1u2/uevent
++ DEVTYPE=wlan
++ INTERFACE=wlp0s29f7u1u2
++ IFINDEX=11
+ '[' wlan = wlan -o -e /sys/class/net/wlp0s29f7u1u2/wireless ']'
+ ifaces=' -iwlp0s29f7u1u2'
+ for i in $(cd /sys/class/net && echo *)
+ DEVTYPE=
+ source /sys/class/net/wlp3s0b1/uevent
++ DEVTYPE=wlan
++ INTERFACE=wlp3s0b1
++ IFINDEX=3
+ '[' wlan = wlan -o -e /sys/class/net/wlp3s0b1/wireless ']'
+ ifaces=' -iwlp0s29f7u1u2 -N -iwlp3s0b1'
+ exec wpa_supplicant -s -u -Dnl80211,wext -c /etc/wpa_supplicant.conf -iwlp0s29f7u1u2 -N -iwlp3s0b1
wpa_supplicant v2.8
Copyright (c) 2003-2019, Jouni Malinen <j@w1.fi> and contributors

This software may be distributed under the terms of the BSD license.
See README for more details.

This product includes software developed by the OpenSSL Project
for use in the OpenSSL Toolkit (http://www.openssl.org/)

usage:
  wpa_supplicant [-BddhKLqqstuvW] [-P<pid file>] [-g<global ctrl>] \
        [-G<group>] \
        -i<ifname> -c<config file> [-C<ctrl>] [-D<driver>] [-p<driver_param>] \
        [-b<br_ifname>] [-e<entropy file>] [-f<debug file>] \
        [-o<override driver>] [-O<override ctrl>] \
        [-N -i<ifname> -c<conf> [-C<ctrl>] [-D<driver>] \
        [-m<P2P Device config file>] \
        [-p<driver_param>] [-b<br_ifname>] [-I<config file>] ...]

drivers:
  nl80211 = Linux nl80211/cfg80211
  wext = Linux wireless extensions (generic)
  wired = Wired Ethernet driver
options:
  -b = optional bridge interface name
  -B = run daemon in the background
  -c = Configuration file
  -C = ctrl_interface parameter (only used if -c is not)
  -d = increase debugging verbosity (-dd even more)
  -D = driver name (can be multiple drivers: nl80211,wext)
  -e = entropy file
  -f = log output to debug file instead of stdout
  -g = global ctrl_interface
  -G = global ctrl_interface group
  -h = show this help text
  -i = interface name
  -I = additional configuration file
  -K = include keys (passwords, etc.) in debug output
  -L = show license (BSD)
  -m = Configuration file for the P2P Device interface
  -N = start describing new interface
  -o = override driver parameter for new interfaces
  -O = override ctrl_interface parameter for new interfaces
  -p = driver parameters
  -P = PID file
  -q = decrease debugging verbosity (-qq even less)
  -s = log output to syslog instead of stdout
  -t = include timestamp in debug messages
  -u = enable DBus control interface
  -v = show version
  -W = wait for a control interface monitor before starting
example:
  wpa_supplicant -Dnl80211 -iwlan0 -c/etc/wpa_supplicant.conf
```

</details>

### Steps to reproduce
1. Set `networking.wireless.enable = true;`
2. Boot with at least 2 WiFi interfaces

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.40, NixOS, 19.03.172500.aade6ded796 (Koi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2.2`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

cc @globin module maintainer

**Comment by lopsided98** [comment]
I just ran into this. wpa_supplicant needs a separate config file for each interface. I think the module should be restructured so that networks are per interface. I need this to work, so I'm going to hack something together, and then make a PR where we can figure out backward compatibility.

**Comment by lheckemann** [comment]
@lopsided98 we already have a module for separate configs per interface, its options are at `networking.supplicant`. For just `networking.wireless`, I think it's ok to just reuse the same config file for each interface.

**Comment by lopsided98** [comment]
Thanks. It is quite unexpected that there are two separate modules with overlapping but slightly different features sets. It should definitely be better documented.

**Comment by globin** [comment]
They should be merged, introducing `networking.supplicant` was a mistake in my opinion.

**Comment by lheckemann** [comment]
I've opened PRs to fix this on master and 19.09.
